lib/internal/source_map.js

```diff
@@ -0,0 +1,122 @@
+const { dirname, resolve } = require('path');
+const fs = require('fs');
+const {
+  normalizeReferrerURL,
+} = require('internal/modules/cjs/helpers');
+const { JSON, Object } = primordials;
+const sourceMapCache = new Map();
+const { fileURLToPath, URL } = require('url');
+
+function maybeCacheSourceMap(filename, content) {
+  if (!process.env.NODE_V8_COVERAGE) return;
+
+  let basePath;
+  try {
+    filename = normalizeReferrerURL(filename);
+    basePath = dirname(fileURLToPath(filename));
+  } catch (err) {
+    // This is most likely an [eval]-wrapper, which is currently not
+    // supported.
+    debug(err.message);
+    return;
+  }
+
+  const match = content.match(/\/[*/]#\s+sourceMappingURL=(?<sourceMappingURL>[^\s]+)/);
+  if (match) {
+    sourceMapCache.set(filename, {
+      url: match.groups.sourceMappingURL,
+      data: dataFromUrl(basePath, match.groups.sourceMappingURL)
+    });
+  }
+}
+
+function dataFromUrl(basePath, sourceMappingURL) {
+  try {
+    const url = new URL(sourceMappingURL);
+    switch (url.protocol) {
+      case 'data:':
+        return sourceMapFromDataUrl(basePath, url.pathname);
+      default:
+        debug(`unknown protocol ${url.protocol}`);
+        return null;
+    }
+  } catch (err) {
+    debug(err.message);
+    // If no scheme is present, we assume we are dealing with a file path:
+    const sourceMapFile = resolve(basePath, sourceMappingURL);
+    return sourceMapFromFile(sourceMapFile);
+  }
+}
+
+function sourceMapFromFile(sourceMapFile) {
+  try {
+    const content = fs.readFileSync(sourceMapFile, 'utf8');
+    const data = JSON.parse(content);
+    return sourcesToAbsolute(dirname(sourceMapFile), data);
+  } catch (err) {
+    debug(err);
+    return null;
+  }
+}
+
+// data:[<mediatype>][;base64],<data> see:
+// https://tools.ietf.org/html/rfc2397#section-2
+function sourceMapFromDataUrl(basePath, url) {
+  const [format, data] = url.split(',');
+  const splitFormat = format.split(';');
+  const contentType = splitFormat[0];
+  const base64 = splitFormat[splitFormat.length - 1] === 'base64';
+  if (contentType === 'application/json') {
+    const decodedData = base64 ?
+      Buffer.from(data, 'base64').toString('utf8') : data;
+    try {
+      const parsedData = JSON.parse(decodedData);
+      return sourcesToAbsolute(basePath, parsedData);
+    } catch (err) {
+      debug(err);
+      return null;
+    }
+  } else {
+    debug(`unknown content-type ${contentType}`);
+    return null;
+  }
+}
+
+// If the sources are not absolute URLs after prepending of the "sourceRoot",
+// the sources are resolved relative to the SourceMap (like resolving script
+// src in a html document).
+function sourcesToAbsolute(base, data) {
+  data.sources = data.sources.map((source) => {
+    source = (data.sourceRoot || '') + source;
+    if (!/^[\\/]/.test(source[0])) {
+      source = resolve(base, source);
+    }
+    if (!source.startsWith('file://')) source = `file://${source}`;
```

Comment:
Does that mean that `source` is always a relative URL then? I.e. no need to URL-encode anything or take care of backtracking via `..`?

Reply:
The specification is a little vague; I don't believe they'd ever have to be URL encoded, as for handling `..`, the spec simply states:

```
If the sources are not absolute URLs after prepending of the “sourceRoot”, the sources are resolved relative to the SourceMap (like resolving script src in a html document).
```

From my testing, I've tried `babel`, `TypeScript`, and `uglify`, which all generate source-maps, it seems like the `resolve(base, source)` logic does the trick. My reasoning for prepending the `file://` and making the path absolute, was for consistency with the module cache, and it means the consumer doesn't need to know what the `cwd` was when the source map was laded.